multicollector.go

```diff
@@ -19,42 +20,88 @@ type multiCollector struct {
 	// Impose a timeout on collection if non-zero.
 	timeout time.Duration
 
+	logger *log.Logger
+
+	warningsDesc *prometheus.Desc
+
 	members []multiCollectorMember
 }
 
 var _ prometheus.Collector = (*multiCollector)(nil)
 
 func newMultiCollector(m ...multiCollectorMember) *multiCollector {
 	return &multiCollector{
+		logger: log.Default(),
+		warningsDesc: prometheus.NewDesc("paperless_warnings_total",
+			"Number of warnings generated while scraping metrics.",
+			nil, nil),
 		members: m,
 	}
 }
 
 func (c *multiCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.warningsDesc
+
 	for _, i := range c.members {
 		i.describe(ch)
 	}
 }
 
-func (c *multiCollector) Collect(ch chan<- prometheus.Metric) {
-	ctx := context.Background()
+func (c *multiCollector) collect(ctx context.Context, ch chan<- prometheus.Metric) error {
+	var wg sync.WaitGroup
 
-	if c.timeout != 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, c.timeout)
-		defer cancel()
-	}
+	collected := make(chan prometheus.Metric)
+
+	defer func() {
+		close(collected)
+
+		wg.Wait()
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		var warnings []error
+
+		for m := range collected {
+			if warning, ok := m.(*warning); ok && warning != nil {
+				warnings = append(warnings, warning.err)
+				continue
+			}
+
+			ch <- m
+		}
+
+		if len(warnings) > 0 {
+			c.logger.Printf("Metrics collection warnings: %q", warnings)
+		}
+
+		ch <- prometheus.MustNewConstMetric(c.warningsDesc, prometheus.GaugeValue, float64(len(warnings)))
```

Comment:
I was wondering if it is possible to add an "error code" or something as a label. Aka indicating WHAT went wrong.
If this metric spikes, you need to watch the logs of the exporter. In a good monitoring system, the metric provides you an indicator on what went wrong.
This metric for now indicates that something went wrong, but not what. It starts the indication.

What do you think about adding a label like error=`remote_verison_scrape_failed`? (other collectors can run other errors).

Reply:
Adding a warning category is indeed a good idea. We shouldn't return the whole message as that would potentially blow up the metric cardinality. I'll do this as a follow-up change.

Reply:
PR #70.